src/type_checker/checker.rs

```diff
@@ -682,6 +675,14 @@ impl<B: Backend> TypeChecker<B> {
 
         // compare argument types with the function signature
         for (sig_arg, (typ, span)) in expected.iter().zip(observed) {
+            // when const attribute presented, the argument must be a constant
+            if sig_arg.is_constant() && !matches!(typ, TyKind::Field { constant: true }) {
+                return Err(self.error(
+                    ErrorKind::ArgumentTypeMismatch(sig_arg.typ.kind.clone(), typ),
+                    span,
+                ));
+            }
```

Comment:
![Screenshot 2024-08-23 at 3 18 57 PM](https://github.com/user-attachments/assets/1a8b560f-cab9-4bd4-a903-1e04e8731252)

^ I managed to write this and it compiles :D perhaps we should add a check that we can only have `const Field` arguments in a function, and not `const CustomType`. If we want to generalize it one day we can do it, but at least we don't have to do the work now, wdyt?

Reply:
it's also weird that now we have `arg.is_constant()` which checks the `FnArg.attribute`. Do we even need this to track constants now? It seems like the `match_expected` call right below this comment would be enough (if you address my earlier comment)

Reply:
my recommendation would be to keep both checks at this point, but add a comment that this might be unnecessary :D

Reply:
I think the `is_constant` is still necessary. This is because for the function signature
- with const attribute, it only accept constant field.
- without const attribute, it accept the field type in either mode.

As I mentioned above, it may be better to create a specific checking rule for function signature, so as to make the rules clearer for different cases and avoid being confused by the `match_expected` function.

Reply:
updated to forbid const over types other than field type